ui/job-view/Push.jsx

```diff
@@ -21,10 +21,65 @@ export default class Push extends React.Component {
     this.hideRunnableJobs = this.hideRunnableJobs.bind(this);
 
     this.state = {
-      runnableVisible: false
+      runnableVisible: false,
+      watched: "none",
+
+      // props.push isn't actually immutable due to the way it hooks up to angular, therefore we
+      // need to keep the previous value in the state.
+      last_job_counts: props.push.job_counts ? Object.assign({}, props.push.job_counts) : null,
     };
   }
 
+  componentWillReceiveProps(nextProps) {
+    this.showUpdateNotifications(nextProps);
+  }
+
+  showUpdateNotifications(nextProps) {
+    if (Notification.permission !== "granted" || this.state.watched === "none") {
+      return;
+    }
+
+    const nextCounts = nextProps.push.job_counts;
+    if (this.state.last_job_counts) {
+      const nextUncompleted = nextCounts.pending + nextCounts.running;
+      const lastUncompleted = this.state.last_job_counts.pending + this.state.last_job_counts.running;
+
+      const nextCompleted = nextCounts.completed;
+      const lastCompleted = this.state.last_job_counts.completed;
+
+      let message;
+      if (lastUncompleted > 0 && nextUncompleted === 0) {
+        message = "Push completed";
+        this.setState({ watched: "none" });
+      } else if (this.state.watched === "job" && lastCompleted < nextCompleted) {
+        const completeCount = nextCompleted - lastCompleted;
+        message = completeCount + " jobs completed";
+      }
+
+      if (message) {
+        const notification = new Notification(message, {
+          body: `${this.props.repoName} rev ${this.props.push.revision.substring(0, 12)}`,
+          tag: this.props.push.id
+        });
+
+        notification.onerror = (event) => {
+          this.thNotify.send(`${event.target.title}: ${event.target.body}`, "danger");
+        };
+
+        notification.onclick = (event) => {
+          if (this.container) {
+            this.container.scrollIntoView();
+            event.target.close();
+          }
+        };
+      }
+    }
+
+    if (nextCounts) {
+      this.setState({ last_job_counts: Object.assign({}, nextCounts) });
```

Comment:
One thing to consider, right now if you ignore the notification for a push while getting per-job notifications, it is replaced by a new notification. The number of newly completed jobs doesn't accumulate, it just counts since the last update. What we could do here is make them accumulate until the user actively clicks on the notification.

Reply:
I'll leave this as is, I think it is fine.